### PR TITLE
Fixing readfile

### PR DESCRIPTION
**Pull request by ronykoz**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/16043

## Description
Fixing a bug where parsing unicode fails

### CODE REVIEW

**Comment by Itay4** [comment]
we should let the user decide the codec with an argument, which its default will be `utf-8'`
also need to add to the test special characters

**Comment by yardensade** [comment]
Should create a test for it with utf-8 chars and non utf-8 chars.

File: Scripts/script-ReadFile.yml [hunk 1]
```diff
@@ -10,7 +10,14 @@ script: |-
       maxFileSize = 1024**2
   res = demisto.executeCommand('getFilePath', {'id': demisto.args()['entryID']})
   with open(res[0]['Contents']['path'], 'r') as f:
-      data = f.read(maxFileSize).decode('unicode_escape').encode('utf-8')
+      data = f.read(maxFileSize)
+      try:
+          data = data.decode('unicode_escape')
```
Comment (glicht) [inline Scripts/script-ReadFile.yml]: I am not clear about why we are doing decode with 'unicode_escape'. Most files we will want to read will be encoded in utf-8 and not unicode_escape. 
```suggestion
          data = data.decode(demisto.args().get("encoding"), 'replace')
```

File: Scripts/script-ReadFile.yml [hunk 2]
```diff
@@ -36,7 +43,11 @@ args:
   required: false
   default: false
   description: Maximal file size to load, in bytes. Default is 1MB.
+- name: encoding
+  description: The encoding type you want for your file
```
Comment (glicht) [inline Scripts/script-ReadFile.yml]:
```suggestion
  description: The encoding type of the file
```

File: Scripts/script-ReadFile.yml [hunk 3]
```diff
@@ -10,7 +10,14 @@ script: |-
...
   res = demisto.executeCommand('getFilePath', {'id': demisto.args()['entryID']})
   with open(res[0]['Contents']['path'], 'r') as f:
-      data = f.read(maxFileSize).decode('unicode_escape').encode('utf-8')
+      data = f.read(maxFileSize)
+      try:
+          data = data.decode('unicode_escape')
+      except UnicodeDecodeError:
+          pass  # Passing in a case we have \U or \N in the file
+
+      data = data.encode(demisto.args().get("encoding"))
```
Comment (glicht) [inline Scripts/script-ReadFile.yml]: data returned to the context should be utf-8 encoded always
```suggestion
      data = data.encode('utf-8')
```

**Comment by ronykoz** [comment]
@glicht is it good to go?

File: Scripts/script-ReadFile.yml [hunk 4]
```diff
@@ -36,7 +39,11 @@ args:
   required: false
   default: false
   description: Maximal file size to load, in bytes. Default is 1MB.
+- name: decode_type
```
Comment (glicht) [inline Scripts/script-ReadFile.yml]: Let's change the name to: encoding
```suggestion
- name: encoding
```

File: Scripts/script-ReadFile.yml [hunk 5]
```diff
@@ -36,7 +39,11 @@ args:
...
+  description: The decoding type of the file
```
Comment (glicht) [inline Scripts/script-ReadFile.yml]:
```suggestion
  description: The character encoding of the file
```

File: Scripts/script-ReadFile.yml [hunk 6]
```diff
@@ -10,7 +10,10 @@ script: |-
       maxFileSize = 1024**2
   res = demisto.executeCommand('getFilePath', {'id': demisto.args()['entryID']})
   with open(res[0]['Contents']['path'], 'r') as f:
-      data = f.read(maxFileSize).decode('unicode_escape').encode('utf-8')
```
Comment (glicht) [inline Scripts/script-ReadFile.yml]: Use: `errors='replace'` for the decode function

Comment (ronykoz) [reply Scripts/script-ReadFile.yml]: What do you mean by `errors='replace'`?

Comment (glicht) [reply Scripts/script-ReadFile.yml]: `decode` method takes as second parameter an `errors` argument. It is best to use `replace` for the `errors` arg.

**Pull request merged**